clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -1116,58 +1116,62 @@ castStringLiteralToDefaultAddressSpace(CIRGenModule &CGM, mlir::StringAttr GV) {
-  Name = Out.str();
-  StringLiteralCnt++;
+  mlir::cir::GlobalOp GV;
+  if (!getLangOpts().WritableStrings && ConstantStringMap.count(C)) {
+    GV = ConstantStringMap[C];
+    // The bigger alignment always wins.
+    if (!GV.getAlignment() ||
+        uint64_t(Alignment.getQuantity()) > *GV.getAlignment())
+      GV.setAlignmentAttr(getSize(Alignment));
+  } else {
+    SmallString<256> StringNameBuffer = Name;
+    llvm::raw_svector_ostream Out(StringNameBuffer);
+    if (StringLiteralCnt)
+      Out << StringLiteralCnt;
+    Name = Out.str();
+    StringLiteralCnt++;
+
+    SmallString<256> MangledNameBuffer;
+    StringRef GlobalVariableName;
+    auto LT = mlir::cir::GlobalLinkageKind::ExternalLinkage;
+
+    // Mangle the string literal if that's how the ABI merges duplicate strings.
+    // Don't do it if they are writable, since we don't want writes in one TU to
+    // affect strings in another.
+    if (getCXXABI().getMangleContext().shouldMangleStringLiteral(S) &&
+        !getLangOpts().WritableStrings) {
+      assert(0 && "not implemented");
+    } else {
+      LT = mlir::cir::GlobalLinkageKind::InternalLinkage;
+      GlobalVariableName = Name;
+    }
 
-  SmallString<256> MangledNameBuffer;
-  StringRef GlobalVariableName;
-  auto LT = mlir::cir::GlobalLinkageKind::ExternalLinkage;
+    auto loc = getLoc(S->getSourceRange());
+    auto typedC = llvm::dyn_cast<mlir::TypedAttr>(C);
+    if (!typedC)
+      llvm_unreachable("this should never be untyped at this point");
+    GV = generateStringLiteral(loc, typedC, LT, *this, GlobalVariableName,
+                               Alignment);
+    ConstantStringMap[C] = GV;
 
-  // Mangle the string literal if that's how the ABI merges duplicate strings.
-  // Don't do it if they are writable, since we don't want writes in one TU to
-  // affect strings in another.
-  if (getCXXABI().getMangleContext().shouldMangleStringLiteral(S) &&
-      !getLangOpts().WritableStrings) {
-    assert(0 && "not implemented");
-  } else {
-    LT = mlir::cir::GlobalLinkageKind::InternalLinkage;
-    GlobalVariableName = Name;
+    assert(!cir::UnimplementedFeature::reportGlobalToASan() && "NYI");
   }
 
-  auto loc = getLoc(S->getSourceRange());
-  auto typedC = llvm::dyn_cast<mlir::TypedAttr>(C);
-  if (!typedC)
-    llvm_unreachable("this should never be untyped at this point");
-  auto GV = generateStringLiteral(loc, typedC, LT, *this, GlobalVariableName,
-                                  Alignment);
-  ConstantStringMap[C] = GV;
+  auto ArrayTy = GV.getSymType().dyn_cast<mlir::cir::ArrayType>();
+  assert(ArrayTy && "String literal must be array");
+  auto PtrTy =
+      mlir::cir::PointerType::get(builder.getContext(), ArrayTy.getEltType());
 
-  assert(!cir::UnimplementedFeature::reportGlobalToASan() && "NYI");
-  return mlir::SymbolRefAttr::get(
-      castStringLiteralToDefaultAddressSpace(*this, GV.getSymNameAttr()));
+  SmallVector<mlir::Attribute, 1> Indices{builder.getI64IntegerAttr(0)};
```

Comment:
We should probably canonicalize this better, but you don't need to pass anything here for `Indices`, which would mean the same as accessing `[0]`. We should unify "empty list" and "[0]". For now if you don't pass this you get more terse CIR output, which is preferable, see next comment.

Reply:
I can do this but don't we want to distinguish these two global views?
```
  cir.global external @a = #cir.global_view<@s> : !cir.ptr<!cir.array<!s8i x 2>> loc(#loc8)
  cir.global external @b = #cir.global_view<@s, [0]> : !cir.ptr<!s8i> loc(#loc9)                                                                                                                           
```

Reply:
From `GlobalViewAttr` description:

>  The type of the input symbol can be different from `#cir.global_view`  output type, since a given view of the global might require a static cast for initializing other globals.

We also embedded a static cast as part of `#cir.global_view`. Note that the indices are not applied to the resulting type...

>  A list of indices can be optionally passed and each element subsequently  indexes underlying types. For `symbol` types like `!cir.array`  and `!cir.struct`, it leads to the constant address of sub-elements, while  for `!cir.ptr`, an offset is applied. The first index is relative to the original symbol type, not the produced one.

... but on the input type of the referred global.

So, back to your example, both `@a` and `@b` point to the beginning of `@s`, even though they output different types. Overall though, I think you can tackle this in another PR though, let's land this for now!

Reply:
> We also embedded a static cast as part of `#cir.global_view`.

I wonder if we may in some situations need a bitcast to align type of `AddressOfOp`/`GEP` to `GlobalView`'s type. E.g. `AddressOfOp` return type must match the type of the global:
```
loc(fused["tmp.c":8:1, "tmp.c":8:17]): error: 'llvm.mlir.addressof' op the type must be a pointer to the type of the referenced global
```

> So, back to your example, both `@a` and `@b` point to the beginning of `@s`, even though they output different types. I believe you can tackle this in another PR though, let's land this for now!

Got it. Let me meditate a bit more on this to ensure I'm not making things worse with my PR.

Reply:
> > We also embedded a static cast as part of `#cir.global_view`.
> 
> I wonder if we may in some situations need a bitcast to align type of `AddressOfOp`/`GEP` to `GlobalView`'s type. E.g. `AddressOfOp` return type must strictly match the type of the global:
> 
> ```
> loc(fused["tmp.c":8:1, "tmp.c":8:17]): error: 'llvm.mlir.addressof' op the type must be a pointer to the type of the referenced global
> ```

If that's the case then the lowering rewriter or some other pass that runs before should be able to introduce the dialect specific needed cast as part of lowering! For LLVM lowering we sometimes to that as part of LLVM lowering or during LoweringPrepare.

> > So, back to your example, both `@a` and `@b` point to the beginning of `@s`, even though they output different types. I believe you can tackle this in another PR though, let's land this for now!
> 
> Got it. Let me meditate a bit more on this to ensure I'm not making things worse with my PR.

Sure! Note that for this PR you could pass an empty attribute so you don't create any indices, and then in a follow up PR you could improve along the lines of what we previously discussed.